### PR TITLE
Chainid network configs

### PR DESCRIPTION
**Pull request by shanejonas**
## Explanation
Adding chainId to the network configurations allows other code to treat the configurations as the same wether they are infura or a custom network.

As a result it simplifies the implementation for `findNetworkClientIdByChainId`.

We need `findNetworkClientIdByChainId` because the dapp API accepts chainId via `switchEthereumChain` and possibly caip-27 in the future. We need a way to map from chainId to networkClientId. If there are multiple networkClientIds for a given chainId we just naively return the first one. 

## References

* Related to https://github.com/MetaMask/MetaMask-planning/issues/1006

### `@metamask/network-controller`

- **ADDED**: `findNetworkClientIdByChainId`
- **ADDED**: chainId to built-in network configurations

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate

### CODE REVIEW

File: packages/network-controller/tests/NetworkController.test.ts [hunk 1]
```diff
@@ -1698,7 +1797,7 @@ describe('NetworkController', () => {
                     'AAAA-AAAA-AAAA-AAAA',
                     {
                       type: NetworkClientType.Custom,
-                      chainId: toHex(1),
+                      chainId: '0x1',
```
Comment (BelfordZ) [inline packages/network-controller/tests/NetworkController.test.ts:1800]: Should revert this change

```suggestion
                      chainId: toHex(1),
```

File: packages/network-controller/tests/NetworkController.test.ts [hunk 2]
```diff
@@ -1790,7 +1901,6 @@ describe('NetworkController', () => {
                     return networkClientId1.localeCompare(networkClientId2);
                   },
                 );
-
```
Comment (BelfordZ) [inline packages/network-controller/tests/NetworkController.test.ts:1793]: revert

File: packages/network-controller/src/NetworkController.ts [hunk 3]
```diff
@@ -1231,6 +1231,25 @@ export class NetworkController extends BaseControllerV2<
...
+      (networkClient) => {
+        return networkClient.configuration.chainId === chainId;
+      },
+    );
+    if (networkConfigurationIndex === -1) {
+      throw new Error("Couldn't find networkClientId for chainId");
+    }
+    return Object.keys(networkClients)[networkConfigurationIndex];
+  }
+
```
Comment (jiexi) [inline packages/network-controller/src/NetworkController.ts:1252]: does Object.values and Object.keys guarantee elements ordered the same? Would using `Object.entries()` be worth it?

File: packages/network-controller/src/NetworkController.ts [hunk 4]
```diff
@@ -1231,6 +1231,25 @@ export class NetworkController extends BaseControllerV2<
     });
   }
 
+  /**
+   * Searches for a network configuration ID with the given ChainID and returns it.
+   *
+   * @param chainId - ChainId to search for
+   * @returns networkClientId of the network configuration with the given chainId
+   */
+  findNetworkClientIdByChainId(chainId: Hex): NetworkClientId {
```
Comment (adonesky1) [inline packages/network-controller/src/NetworkController.ts:1240]: Wondering if it would be more future proof for the us to return all the networkClients for the chainId and let the consumer decide what to do if there are more than one?

Comment (adonesky1) [reply packages/network-controller/src/NetworkController.ts:1240]: Ok fair enough!

**Review by jiexi — APPROVED** [review]

**Review by adonesky1 — DISMISSED** [review]

File: packages/network-controller/src/NetworkController.ts [hunk 5]
```diff
@@ -1302,6 +1319,7 @@ export class NetworkController extends BaseControllerV2<
         type: NetworkClientType.Infura,
         network,
         infuraProjectId: this.#infuraProjectId,
+        chainId: BUILT_IN_NETWORKS[network].chainId,
```
Comment (adonesky1) [inline packages/network-controller/src/NetworkController.ts:1322]: Is it potentially a problem that `BUILT_IN_NETWORKS` includes `NetworkType.rpc`?

Comment (BelfordZ) [reply packages/network-controller/src/NetworkController.ts:1322]: not AFAIK

Comment (adonesky1) [reply packages/network-controller/src/NetworkController.ts:1322]: oh wait nvm

**Review by adonesky1 — APPROVED** [review]
LGTM!

File: packages/network-controller/src/types.ts [hunk 6]
```diff
@@ -30,6 +30,7 @@ export type CustomNetworkClientConfiguration = {
  * network.
  */
 export type InfuraNetworkClientConfiguration = {
+  chainId: Hex;
```
Comment (mcmire) [inline packages/network-controller/src/types.ts:33]: Hmm... this doesn't seem like the right approach to me.

First, the purpose of the network client configuration is to configure the network client. Currently, we don't need this option; we can use the Infura network name to figure out the appropriate chain ID. Therefore, adding this property expands the responsibility of this type unnecessarily.

Second, this property makes this type ineffective. With this property added, it is now possible to create an InfuraNetworkClientConfiguration that looks like this:

``` typescript
{
  type: NetworkClientType.Infura,
  chainId: "0x1337",
  network: InfuraNetworkType.mainnet,
  infuraProjectId: "infura-project-id"
}
```

That doesn't look right to me. The chain ID associated with an Infura network should never be different than we know it to be and how we construct the network client. If it were, then we'd have a custom network, not an Infura network.

One of the goals of types is to make it impossible for the system to get into an invalid state (even if it is just an accident). This change seems to not do that.

I don't think this property is necessary. I understand that one of our goals is to consolidate the definitions of Infura and custom network clients, but I feel that there is a more cohesive approach we can take.

In this case, we already know what the chain ID for an Infura network is, because we can get it from `BUILT_IN_NETWORKS`. I think a better approach would have been to create a new network client configuration object that combines a custom and an Infura one together so that it now has a common `chainId` and `rpcUrl`. Then converting an Infura network configuration into this common form would look like:

```
{
  chainId: BUILT_IN_NETWORKS.mainnet.chainId,
  rpcUrl: buildInfuraRpcUrl(InfuraNetworkType.mainnet, 'infura-project-id')
}
```

Comment (mcmire) [reply packages/network-controller/src/types.ts:33]: An update: @adonesky1, @BelfordZ and @shanejonas and I talked about this. They provided more context behind this change. We need a way to access the chain ID for a network client across many many controllers that we are making multichain capable and this just seemed like the most direct way of solving that constraint without jumping through a bunch of hoops.

Zooming out, this is another case in which the Infura/custom network split is causing difficulties. So of course we know that consolidating the two would solve a lot of problems.

As a more immediate fix, though, we can remove `network` from InfuraNetworkClientConfiguration in favor of `chainId`. In the same way that we can derive chain ID from Infura network name, we can also do the reverse. This should make it possible to access the chain ID of an Infura network client while still preserving the type safety of the two network clients.
